contracts/contracts/lib/LibGateway.sol

```diff
@@ -355,6 +380,19 @@ library LibGateway {
     /// @param arrivingFrom - the immediate subnet from which this message is arriving
     /// @param crossMsg - the cross message to be executed
     function applyMsg(SubnetID memory arrivingFrom, IpcEnvelope memory crossMsg) internal {
```

Comment:
this method seems confusing though, because it's called `applyMsg` but actually it's expecting bottom up only.

Reply:
It does not expect only bottom-up messages; it expects both. But I understand it can be confusing. :)